daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/SequenceChild.scala

```diff
@@ -27,216 +27,434 @@ import org.apache.daffodil.schema.annotation.props.gen.OccursCountKind
+   * including the behavior for daf:emptyElementPolicy 'emptySuppressed' which special cases
+   * Required elements like scalars, iff they are emptyRep, emptyValueDelimiterPolicy,
+   * nilValueDelimiterPolicy, complex elements that nillable, or fully defaultable.
+   *
+   * So we have ((simpleStringHexBinary x (emptySuppressed, emptyAllowed), simpleOther, complex) x (nillable, not) x
+   * 4 behaviors. That's 32 combinations. Let's start with fewer cases and more runtime
+   * decisions, and specialize if we think it will help clarity or performance.
+   */
+  private lazy val sepScalarElementHelper: SeparatedHelper = {
+    Assert.invariant(e.isScalar)
+    Assert.invariant(e.isRepresented)
+    val isd = e.isSimpleType && (e.lengthKind eq LengthKind.Delimited)
+    this.sscb match {
+      case PositionalTrailingLax |
+        PositionalTrailingStrict => new PositionalTrailingScalarElementSeparatedSequenceChildParseResultHelper(sscb, erd, isd, eep, isERZL, isERNZL)
+      case Positional => new PositionalScalarElementSeparatedSequenceChildParseResultHelper(sscb, erd, isd, eep, isERZL, isERNZL)
+      case NonPositional => new NonPositionalScalarElementSeparatedSequenceChildParseResultHelper(sscb, erd, isd, eep, isERZL, isERNZL)
+    }
+  }
+
+  override lazy val unseparatedHelper = {
+    term match {
+      case _: ModelGroup => unsepGroupHelper
+      case _: ElementBase => unsepScalarElementHelper
+    }
+  }
+
+  private lazy val unsepGroupHelper: UnseparatedHelper = {
+    new GroupUnseparatedSequenceChildParseResultHelper(mgrd, isModelGroupRepPossiblyZeroLength, isModelGroupRepNonZeroLength)
+  }
+
+  private lazy val unsepScalarElementHelper: UnseparatedHelper = {
+    new ScalarElementUnseparatedSequenceChildParseResultHelper(erd, eep, isEmptyRepZeroLength, isEmptyRepNonZeroLength)
   }
 }
 
-sealed abstract class ElementSequenceChild(
-  sq: SequenceTermBase,
+sealed abstract class RepElementSequenceChild(sq: SequenceTermBase,
   protected val e: ElementBase, groupIndex: Int)
   extends SequenceChild(sq, e, groupIndex) {
+  import SeparatedSequenceChildBehavior._
 
-  protected lazy val erd = e.elementRuntimeData
-}
+  Assert.usage(!e.isScalar)
 
-sealed trait RepUnparserMixin { self: ElementSequenceChild =>
+  lazy val sequenceChildUnparser: SequenceChildUnparser =
+    sq.hasSeparator match {
+      case true => {
+        new RepOrderedSeparatedSequenceChildUnparser(
+        childUnparser, srd, erd, sepUnparser, sq.separatorPosition, sq.separatorSuppressionPolicy,
+          e.zeroLengthDetector, e.isPotentiallyTrailing,
+          isKnownStaticallyNotToSuppressSeparator,
+          isPositional,
+          isDeclaredLast)
+      }
+      case false => new RepOrderedUnseparatedSequenceChildUnparser(childUnparser, srd, erd)
+    }
 
-  protected def e: ElementBase
+  private lazy val eep = e.emptyElementPolicy
+
+  protected lazy val separatedHelper: SeparatedHelper = {
```

Comment:
scaladoc reminder. There are two different kinds of helper objects here. This one just takes care of infix/prefix/postifix separator.